File: src/db/cms/Mitarbeiter.ts
import { z } from "zod";
import { client, config } from "../config";

const Mitarbeiter = z.object({
  ID: z.string(),
  Name: z.string(),
  Short: z.string(),
  Sex: z.string(),
  Tags: z.string(),
  Focus: z.string(),
  Abteilungid: z.string(),
  Image: z.boolean(),
});
type Mitarbeiter = z.infer<typeof Mitarbeiter>;

const GetMitarbeiterProps = z.object({ id: z.string() });
type GetMitarbeiterProps = z.infer<typeof GetMitarbeiterProps>;

const UpdateMitarbeiterProps = Mitarbeiter;
type UpdateMitarbeiterProps = z.infer<typeof UpdateMitarbeiterProps>;

const CreateMitarbeiterProps = z.object({
  Name: z.string(),
  Short: z.string(),
  Sex: z.string(),
  Tags: z.string(),
  Focus: z.string(),
  Abteilungid: z.string(),
  Image: z.boolean(),
});
type CreateMitarbeiterProps = z.infer<typeof CreateMitarbeiterProps>;

const getAllMitarbeiter = async (): Promise<Mitarbeiter[] | null> => {
  const res = await client.get<Mitarbeiter[] | { error: string }>(
    `/CMS/Mitarbeiter/`,
    config
  );
  if ("error" in res.data) {
    return null;
  }
  return res.data;
};

const getMitarbeiter = async (
  props: GetMitarbeiterProps
): Promise<Mitarbeiter | null> => {
  const res = await client.get<Mitarbeiter | { error: string }>(
    `/CMS/Mitarbeiter/${props.id}`,
    config
  );
  if ("error" in res.data) {
    return null;
  }
  return res.data;
};

const updateMitarbeiter = async (
  props: UpdateMitarbeiterProps
): Promise<Mitarbeiter | null> => {
  const form = new FormData();
  form.set("ID", props.ID);
  form.set("Name", props.Name);
  form.set("Short", props.Short);
  form.set("Sex", props.Sex);
  form.set("Tags", props.Tags);
  form.set("Focus", props.Focus);
  form.set("Abteilungid", props.Abteilungid);
  form.set("Image", props.Image ? "true" : "false");
  const res = await client.post<Mitarbeiter | { error: string }>(
    `/CMS/Mitarbeiter/${props.ID}/edit`,
    form,
    config
  );
  if ("error" in res.data) {
    return null;
  }
  return res.data;
};

const createMitarbeiter = async (
  props: CreateMitarbeiterProps
): Promise<Mitarbeiter | null> => {
  const form = new FormData();
  form.set("Name", props.Name);
  form.set("Short", props.Short);
  form.set("Sex", props.Sex);
  form.set("Tags", props.Tags);
  form.set("Focus", props.Focus);
  form.set("Abteilungid", props.Abteilungid);
  form.set("Image", props.Image ? "true" : "false");
  const res = await client.post<Mitarbeiter | { error: string }>(
    `/CMS/Mitarbeiter/new`,
    form,
    config
  );
  if ("error" in res.data) return null;
  return res.data;
};

const deleteMitarbeiter = async (props: GetMitarbeiterProps): Promise<void> => {
  const form = new FormData();
  form.set("ID", props.id);
  await client.post<{ error: string }>(
    `/CMS/Mitarbeiter/${props.id}/delete`,
    form,
    config
  );
};

export {
  getAllMitarbeiter,
  getMitarbeiter,
  updateMitarbeiter,
  createMitarbeiter,
  deleteMitarbeiter,
  type Mitarbeiter,
  type GetMitarbeiterProps,
  type CreateMitarbeiterProps,
  type UpdateMitarbeiterProps,
};

File: cmsroutes/mitarbeiter.go
package cmsroutes

import (
	"database/sql"
	"encoding/json"
	"net/http"
	"time"

	"github.com/computerextra/golang-backend/cms"
	"github.com/computerextra/golang-backend/env"
	"github.com/computerextra/golang-backend/helper"
	"github.com/google/uuid"
	"github.com/gorilla/mux"
)

func GetAllMitarbeiter(w http.ResponseWriter, r *http.Request) {
	r.Header.Add("Content-Type", "application/json")
	w.Header().Set("Content-Type", "application/json")
	ctx := r.Context()

	// Get Database
	datebase, err := sql.Open("mysql", env.GetEnv("DATABASE_URL"))
	if err != nil {
		panic(err)
	}
	datebase.SetConnMaxIdleTime(time.Minute * 3)
	datebase.SetMaxOpenConns(10)
	datebase.SetMaxIdleConns(10)
	queries := cms.New(datebase)
	User, err := queries.GetAllMitarbeiter(ctx)
	datebase.Close()

	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
	} else {
		json.NewEncoder(w).Encode(User)
	}
}

func GetMitarbeiter(w http.ResponseWriter, r *http.Request) {
	id := mux.Vars(r)["id"]
	r.Header.Add("Content-Type", "application/json")
	w.Header().Set("Content-Type", "application/json")
	ctx := r.Context()

	// Get Database
	datebase, err := sql.Open("mysql", env.GetEnv("DATABASE_URL"))
	if err != nil {
		panic(err)
	}
	datebase.SetConnMaxIdleTime(time.Minute * 3)
	datebase.SetMaxOpenConns(10)
	datebase.SetMaxIdleConns(10)
	queries := cms.New(datebase)
	User, err := queries.GetMitarbeiter(ctx, id)
	datebase.Close()

	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
	} else {
		json.NewEncoder(w).Encode(User)
	}
}

func UpdateMitarbeiter(w http.ResponseWriter, r *http.Request) {
	Name := r.FormValue("Name")
	ID := r.FormValue("ID")
	Short := r.FormValue("Short")
	Image := r.FormValue("Image")
	Sex := r.FormValue("Sex")
	Tags := r.FormValue("Tags")
	Focus := r.FormValue("Focus")
	Abteilungid := r.FormValue("Abteilungid")
	r.Header.Add("Content-Type", "application/json")
	w.Header().Set("Content-Type", "application/json")
	ctx := r.Context()

	// Get Database
	datebase, err := sql.Open("mysql", env.GetEnv("DATABASE_URL"))
	if err != nil {
		panic(err)
	}
	datebase.SetConnMaxIdleTime(time.Minute * 3)
	datebase.SetMaxOpenConns(10)
	datebase.SetMaxIdleConns(10)
	queries := cms.New(datebase)
	User, err := queries.UpdateMitarbeiter(ctx, cms.UpdateMitarbeiterParams{
		Name:        Name,
		ID:          ID,
		Short:       Short,
		Image:       helper.If(Image == "true", true, false),
		Sex:         Sex,
		Tags:        Tags,
		Focus:       Focus,
		Abteilungid: Abteilungid,
	})
	datebase.Close()

	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
	} else {
		json.NewEncoder(w).Encode(User)
	}
}

func CreateMitarbeiter(w http.ResponseWriter, r *http.Request) {
	Name := r.FormValue("Name")
	Short := r.FormValue("Short")
	Image := r.FormValue("Image")
	Sex := r.FormValue("Sex")
	Tags := r.FormValue("Tags")
	Focus := r.FormValue("Focus")
	Abteilungid := r.FormValue("Abteilungid")
	r.Header.Add("Content-Type", "application/json")
	w.Header().Set("Content-Type", "application/json")
	ctx := r.Context()

	// Get Database
	datebase, err := sql.Open("mysql", env.GetEnv("DATABASE_URL"))
	if err != nil {
		panic(err)
	}
	datebase.SetConnMaxIdleTime(time.Minute * 3)
	datebase.SetMaxOpenConns(10)
	datebase.SetMaxIdleConns(10)
	queries := cms.New(datebase)
	User, err := queries.CreateMitarbeiter(ctx, cms.CreateMitarbeiterParams{
		ID:          uuid.New().String(),
		Name:        Name,
		Short:       Short,
		Image:       helper.If(Image == "true", true, false),
		Sex:         Sex,
		Tags:        Tags,
		Focus:       Focus,
		Abteilungid: Abteilungid,
	})
	datebase.Close()

	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
	} else {
		json.NewEncoder(w).Encode(User)
	}
}

func DeleteMitarbeiter(w http.ResponseWriter, r *http.Request) {
	ID := r.FormValue("ID")
	r.Header.Add("Content-Type", "application/json")
	w.Header().Set("Content-Type", "application/json")
	ctx := r.Context()

	// Get Database
	datebase, err := sql.Open("mysql", env.GetEnv("DATABASE_URL"))
	if err != nil {
		panic(err)
	}
	datebase.SetConnMaxIdleTime(time.Minute * 3)
	datebase.SetMaxOpenConns(10)
	datebase.SetMaxIdleConns(10)
	queries := cms.New(datebase)
	err = queries.DeleteMitarbeiter(ctx, ID)
	datebase.Close()

	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
	} else {
		json.NewEncoder(w).Encode(map[string]string{"error": ""})
	}
}
